Makefile

```diff
@@ -6,66 +6,119 @@
-#    Updated: 2023/11/29 16:17:17 by jodufour         ###   ########.fr        #
+#    Updated: 2024/02/05 22:43:02 by jodufour         ###   ########.fr        #
 #                                                                              #
 # **************************************************************************** #
 
 ######################################
 #              COMMANDS              #
 ######################################
-CXX			=	c++
-LINK		=	clang++
-MKDIR		=	mkdir -p
-RM			=	rm -rf
+   AR = ${shell which ar} rcs
+  CXX = ${shell which c++}
+ LINK = ${shell which clang++}
+MKDIR =	${shell which mkdir} -p
+   RM = ${shell which rm} -rf
 
 ######################################
 #             EXECUTABLE             #
 ######################################
-NAME		=	ft_irc
+NAME = ft_irc
+
+#######################################
+#               LIBRARY               #
+#######################################
+LIB = lib${NAME}.a
 
 #######################################
 #             DIRECTORIES             #
 #######################################
-SRC_DIR		=	src
-OBJ_DIR		=	obj
-PRV_DIR		=	private
+SRC_DIR = src
+LIB_DIR = lib
+OBJ_DIR = obj
+PRV_DIR = private
+INC_DIR = include
 
 ######################################
 #            SOURCE FILES            #
 ######################################
-SRC			=	\
-				main.cpp
+SRC = \
+	${addsuffix .cpp, \
+		main \
+	}
+
+LIB_SRC = \
+	${addsuffix .cpp, \
+		${addprefix ${LIB_DIR}/, \
+			${addprefix command/, \
+				away \
+				die \
+				error \
+				info \
+				invite \
+				ison \
+				join \
+				kick \
+				kill \
+				list \
+				mode \
+				motd \
+				names \
+				nick \
+				notice \
+				oper \
+				part \
+				pass \
+				ping \
+				pong \
+				privmsg \
+				quit \
+				restart \
+				time \
+				topic \
+				user \
+				version \
+				wallops \
+				whois \
+			} \
+		} \
+	}
 
 ######################################
 #            OBJECT FILES            #
 ######################################
-OBJ			=	${SRC:.cpp=.o}
-OBJ			:=	${addprefix ${OBJ_DIR}/, ${OBJ}}
-
-DEP			=	${OBJ:.o=.d}
+    OBJ = ${addprefix ${OBJ_DIR}/,${SRC:.cpp=.o}}
+    DEP = ${OBJ:.o=.d}
+LIB_OBJ = ${addprefix ${OBJ_DIR}/,${LIB_SRC:.cpp=.o}}
+LIB_DEP = ${LIB_OBJ:.o=.d}
 
 #######################################
 #                FLAGS                #
 #######################################
-CXXFLAGS	=	-c
-CXXFLAGS	+=	-Wall -Wextra -Werror
-CXXFLAGS	+=	-MMD -MP
-CXXFLAGS	+=	-Wshadow
-CXXFLAGS	+=	-std=c++98
-CXXFLAGS	+=	-I${PRV_DIR}
+CXXFLAGS = \
+	-c \
+	-Wall -Wextra -Werror \
+	-MMD -MP \
+	-Wshadow \
+	-std=c++98 \
+	-I${PRV_DIR} \
+	-I${INC_DIR}
 
 ifeq (${DEBUG}, 1)
-	CXXFLAGS	+=	-g
-	CXXFLAGS	+=	-DDEBUG=1
+	CXXFLAGS += -g -DDEBUG=1
 endif
 
 #######################################
 #                RULES                #
 #######################################
-${NAME}: ${OBJ}
+.PHONY:	all clean fclean re fre
```

Comment:
Are .PHONY not usually at the end?

Reply:
Even if we were used to putting it at the end of the file, it's a better practice to put it before the rules it contains. This is merely to optimize the `make` performance, winning only a few nanoseconds. To go deeper into the details, when `make` is invoked, it will read the `Makefile` until it finds the wanted target. Then, it will check whether the target is part of the `.PHONY` target. If it does, then `make` will execute the content of the target regardless of whether the file with the same name exists and is up to date. But knowing that `make` reads the `Makefile` only up to the wanted target, if it reaches it without reaching the `.PHONY`, it won't know that the wanted target is part of it. And `make` knows that it didn't read the entire file, so it will re-read it until either the end of the file is reached, or it finds a `.PHONY` target that contains the wanted target.
But now, suppose that the `.PHONY` that contains the wanted target is placed before it. This way, `make` will read the `.PHONY` before reaching the wanted target, and thus, it will already know that it is part of the `.PHONY` target. This will prevent `make` to re-read the `Makefile`.
Now that this is said, there exist 2 communities, those who prefer to have only 1 `.PHONY` target, so they will put it at the beginning only once, and those who prefer to have multiple `.PHONY` targets, so they will put each right before the target that they want to mark as phony. I prefer the first option.